utils/vcsdata.go

```diff
@@ -0,0 +1,147 @@
+package utils
+
+import (
+	"bufio"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
+)
+
+type (
+	VcsDetails struct {
+		vcsRootDirectory sync.Map //key: path to dot git, value: vcsData
+		vcsDirectory     sync.Map //key: path to folder, value: index of vcsRootDirectory
+		vcsDirectorySize *int32   // Size of vcs folders entries
+	}
+	vcsData struct {
+		url      string
+		revision string
+	}
+)
+
+const MAX_ENTRIES = 10000
+
+func NewVcsDetals() *VcsDetails {
+	return &VcsDetails{vcsRootDirectory: sync.Map{}, vcsDirectory: sync.Map{}, vcsDirectorySize: new(int32)}
+}
+
+func (this *VcsDetails) increment(num int32) {
+	atomic.AddInt32(this.vcsDirectorySize, num)
+}
+
+func (this *VcsDetails) get() int32 {
+	return atomic.LoadInt32(this.vcsDirectorySize)
+}
+
+/*
+	Start search for '.git' dir for the current path, incase there is one, extract the details and add a new entry to the cache(hash-map,key:path,value:pointer to the git details).
+	otherwise, search in the parent folder and try:
+	1. search for .git, and save the details for the current dir and all subpath
+	2. .git not found, go to parent dir and repeat
+	3. not found on the root directory, add all subpath to cache with nil as a value
+*/
+func (this *VcsDetails) GetVcsData(path string) (revision, refUrl string, err error) {
+	keys := strings.Split(path, string(os.PathSeparator))
+	var subPath string
+	var subPaths []string
+	var vcsDataResult *vcsData
+	for i := len(keys); i > 0; i-- {
+		subPath = strings.Join(keys[:i], string(os.PathSeparator))
+		// Try to get from cache
+		if searchResult, found := this.searchCache(subPath); found {
+			if data, ok := searchResult.(*vcsData); ok {
```

Comment:
Can we avoid this casting?

Reply:
https://github.com/jfrog/jfrog-client-go/pull/105#discussion_r358798145